third-party/spamassassin/src/main/java/org/apache/james/spamassassin/SpamAssassinInvoker.java

```diff
@@ -111,48 +114,69 @@ public SpamAssassinInvoker(MetricFactory metricFactory, String spamdHost, int sp
-                writer.write(CRLF);
-
-                Arrays.stream(additionalHeaders)
-                    .forEach(header -> {
-                        writer.write(header);
-                        writer.write(CRLF);
-                    });
-
-                writer.write(CRLF);
-                writer.flush();
-
-                // pass the message to spamd
-                message.writeTo(out);
-                out.flush();
-                socket.shutdownOutput();
-
-                SpamAssassinResult spamAssassinResult = in.lines()
-                    .filter(this::isSpam)
-                    .map(this::processSpam)
-                    .findFirst()
-                    .orElse(SpamAssassinResult.empty());
-
-                LOGGER.debug("spam check result: {}", spamAssassinResult);
-                return spamAssassinResult;
-            } catch (UnknownHostException e) {
-                throw new MessagingException("Error communicating with spamd. Unknown host: " + spamdHost);
-            } catch (IOException | MessagingException e) {
-                throw new MessagingException("Error communicating with spamd on " + spamdHost + ":" + spamdPort, e);
-            }
+        Mono<? extends Connection> connectionMono = createConnection();
+
+        return connectionMono.flatMap(connection -> {
+            sendScanMailRequest(message, connection, additionalHeaders);
+            Mono<SpamAssassinResult> result = getSpamAssassinResultMono(connection.inbound());
+            return result;
         });
     }
 
+    private Mono<? extends Connection> createConnection() {
+        return TcpClient.create()
+            .host(spamdHost)
+            .port(spamdPort)
+            .doOnConnected(c -> c.addHandlerLast("codec",
+                new LineBasedFrameDecoder(8 * 1024)))
+            .wiretap(true)
+            .connect();
+    }
+
+    private Mono<SpamAssassinResult> getSpamAssassinResultMono(NettyInbound inbound) {
+        return inbound.receive().asString()
+            .flatMapIterable(s -> Arrays.asList(s.split("\\n")))
+            .filter(this::isSpam)
+            .map(this::processSpam)
+            .singleOrEmpty()
+            .switchIfEmpty(Mono.just(SpamAssassinResult.empty()));
+    }
+
+    private void sendScanMailRequest(MimeMessage message, Connection connection, String[] additionalHeaders) {
+    Mono<byte[]> messageAsBytes =  Mono.fromCallable(() -> {
+              ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+              message.writeTo(byteArrayOutputStream);
+              return byteArrayOutputStream.toByteArray();
+          });
+
+        sendRequestToSpamd(Mono.fromCallable(() -> headerAsString(additionalHeaders)), messageAsBytes, connection);
+    }
+
+    private void sendRequestToSpamd(Mono<String> header, Mono<byte[]> message, Connection connection) {
+        connection.outbound()
+            .sendString(header)
+            .sendByteArray(message)
```

Comment:
Can't we send a Flux<ByteBuffer> we get from an inputStream using ReactorUtils::toChunks ?